### PR TITLE
`_module.specialArgs` + unit test for nixos/documentation module

### LINKED ISSUE #181009: `documentation.nixos.includeAllModules = true` causes infinite recursion
### Describe the bug

`documentation.nixos.includeAllModules = true` causes infinite recursion with specific modules. Specifically importing [this module](https://github.com/pinpox/lollypops/blob/706dc38ccf39b7cbff91977034ecbc1035d213d8/module.nix) and setting the above option results in an infinite recursion. The full stack trace is here: https://gist.github.com/pinpox/edb874416067048a044eae1656207cca

With the kind help of @roberth over on matrix we've narrowed it down to a problem of the manual build when using `specialArgs`:

>  the manual build doesn't include the custom specialArgs

I have tried working around it by setting `buildDocsInSandbox = true` in the module (`false` doesn't work either) and by adding a `defaultText` to each option of the module.

Quoting:

> I think this affects any config that imports from a specialArgs module argument
> the problematic specialArgs is here https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/misc/documentation.nix#L44
> we ought to wire that up somehow. Probably best to add specialArgs as _module.specialArgs in the module system

@roberth let me know if I forgot any useful information. Also pinging @pennae as he might be able to help with this.

### Steps To Reproduce
Steps to reproduce the behaviour:
1. set `documentation.nixos.includeAllModules = true`
2. Import lollypops.nixosModules.lollypops as shown [here](https://github.com/pinpox/nixos/blob/go-task/modules/lollypops-common/default.nix#L4)
3. `nixos-rebuild switch`

### Expected behaviour
Documentation is build without errors

### Additional context
- https://gist.github.com/pinpox/edb874416067048a044eae1656207cca
- https://github.com/pinpox/nixos/blob/go-task/modules/lollypops-common/default.nix#L4
- https://github.com/pinpox/lollypops/blob/706dc38ccf39b7cbff91977034ecbc1035d213d8/module.nix

### Notify maintainers

@roberth @pennae @alyssais @kraem 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.51, NixOS, 22.11 (Raccoon), 22.11.20220708.b39924f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.9.1`
 - channels(root): `""`
 - channels(pinpox): `"home-manager"`
 - nixpkgs: `/nix/store/02pg1kr963gyk2qs3m9hg1lvfgvvhmjy-source`
```

### PR DESCRIPTION
**Pull request by roberth**
###### Description

 - Fixes https://github.com/NixOS/nixpkgs/issues/181009, an infinite recursion.
 - Some pure refactoring of modules
 - New unit test

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
@ofborg test documentation simple

**Comment by SuperSandro2000** [comment]
Is ``nixos/modules/misc/documentation/test-dummy.chapter.xml`` supposed to be empty?

File: lib/modules.nix [hunk 1]
```diff
@@ -266,6 +266,16 @@ rec {
               turned off.
             '';
           };
+
+          _module.specialArgs = mkOption {
```
Comment (infinisil) [inline lib/modules.nix:270]: I don't see this being used in the rest of this PR, is this necessary?

File: nixos/modules/misc/documentation.nix [hunk 2]
```diff
@@ -327,12 +338,14 @@ in
       environment.systemPackages = []
         ++ optional cfg.man.enable manual.manpages
         ++ optionals cfg.doc.enable [ manual.manualHTML nixos-help ];
-
-      services.getty.helpLine = mkIf cfg.doc.enable (
-          "\nRun 'nixos-help' for the NixOS manual."
-      );
     })
 
+    (optionalAttrs (options?services.getty) (
```
Comment (infinisil) [inline nixos/modules/misc/documentation.nix:343]: I don't like `option ? ` checks if they're avoidable. In this case, how about turning it around: Moving this definition into the getty module directly. Could also condition it on getty being enabled.

**Review by infinisil — CHANGES_REQUESTED** [review]

File: lib/modules.nix (hunk 1 above)
Comment (roberth) [reply lib/modules.nix:270]: It sets the appropriate option flags, such as `readOnly`.
It sets the value as well, in `default`. Not sure if we want the default for docs. Maybe not actually. Will move that definition to `config`.

File: nixos/modules/misc/documentation.nix (hunk 2 above)
Comment (roberth) [reply nixos/modules/misc/documentation.nix:343]: I don't like it either actually, but I figured this implementation was less scope-creep-y. I'll move it.

**Pull request merged**